sdk/storage/Azure.Storage.Common/src/GeoRedundantReadPolicy.cs

```diff
@@ -3,54 +3,76 @@
-
-        public override void OnSendingRequest(HttpPipelineMessage message)
+        public GeoRedundantReadPolicy(Uri secondaryStorageUri)
         {
-            if (this._secondaryStorageHost == null)
+            if (secondaryStorageUri == null)
             {
-                return;
+                throw Errors.ArgumentNull(nameof(secondaryStorageUri));
             }
+            this._secondaryStorageHost = secondaryStorageUri.Host;
+        }
+        public override void OnSendingRequest(HttpPipelineMessage message)
+        {
             if (message.Request.Method != RequestMethod.Get && message.Request.Method != RequestMethod.Head)
             {
                 return;
             }
-            var nextHost = message.TryGetProperty(NextHostKey, out var previousHostObj) ? previousHostObj as string : null;
-            if (nextHost == null) // first time being called for the message, so we won't toggle the host
+
+            // Look up what the alternate host is set to in the message properties. For the initial request, this will
+            // not be set.
+            var alternateHost =
+                message.TryGetProperty(
+                    Constants.GeoRedundantReadConstants.AlternateHostKey,
+                    out var alternateHostObj)
+                ? alternateHostObj as string
+                : null;
+            if (alternateHost == null)
             {
-                message.SetProperty(NextHostKey, this._secondaryStorageHost); // queue up the secondary host for subsequent retries
+                // queue up the secondary host for subsequent retries
+                message.SetProperty(Constants.GeoRedundantReadConstants.AlternateHostKey, this._secondaryStorageHost);
                 return;
             }
-            var resourceNotReplicatedToSecondaryStorage = !message.TryGetProperty(ResourceNotReplicatedToSecondaryStorageKey,
-                out var resourceNotReplicatedValue) ? false : (bool)resourceNotReplicatedValue;
-            // if the flag has been set, then we must currently be using the primary storage host and we do not want to allow switching to the secondary host.
-            // As a result, we don't need to worry about the case of getting stuck on the secondary host, so it is safe to just return here.
+
+            // Check the flag that indicates whether the resource has not been propagated to the secondary host yet.
+            // If this flag is set, we don't want to retry against the secondary host again for any subsequent retries.
+            // Also, the flag being set implies that the current request must already be set to the primary host, so we
+            // are safe to return without checking if the current host is secondary or primary.
+            var resourceNotReplicatedToSecondaryStorage =
+                message.TryGetProperty(Constants.GeoRedundantReadConstants.ResourceNotReplicatedToSecondaryStorageKey, out var value)
```

Comment:
Maybe it's worth renaming to `ResourceNotReplicated` so these lines of code end up easier to read?